### PR TITLE
[vi-mode] Supports 'diw' text-object command.

### LINKED ISSUE #2057: [vi-mode] Support diw command
## Description of the new feature/enhancement

As alluded to in #584, PSRL should support vim "text-objects".
Consider supporting the <kbd>d</kbd><kbd>i</kbd><kbd>w</kbd> command to delete inner words.

## Proposed technical implementation details (optional)

I think a new _Chord_ dispatch table must be introduced because currently, PSRL cannot handle 3-letter commands.
Most of the implementation from `ProcessOneKey` can be reused and refactored in common methods.

### PR DESCRIPTION
**Pull request by springcomp**
# PR Summary

Fixes #2057.

This PR is a first attempt at supporting text-objects in vi-mode. It includes a reasonably complete implementation of the <kbd>d</kbd><kbd>i</kbd><kbd>w</kbd> command.

Although I have worked hard to make this PR _production ready_, it must still be considered WIP. I would really appreciate feedback about the implementation.

If this is deemed interesting enough, it could be the first step towards a much complete set of pull requests as alluded to in #584.

## PR Checklist

- [x] PR has a meaningful title
    - Use the present tense and imperative mood when describing your changes
- [x] Summarized changes
- [x] Make sure you've added one or more new tests
- [x] Make sure you've tested these changes in terminals that PowerShell is commonly used in (i.e. conhost.exe, Windows Terminal, Visual Studio Code Integrated Terminal, etc.)
- **User-facing changes**
    - [x] Documentation needed at [PowerShell-Docs](https://github.com/MicrosoftDocs/PowerShell-Docs)
        - [ ] Doc Issue filed: _pending feedback_

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/PowerShell/PSReadLine/pull/2059)

### CODE REVIEW

**Comment by daxian-dbw** [comment]
@springcomp I resolved the conflicts. Will finish the review tonight.

File: PSReadLine/StringBuilderLinewiseExtensions.cs [hunk 1]
```diff
@@ -72,6 +72,23 @@ internal static Range GetRange(this StringBuilder buffer, int lineIndex, int lin
...
+        /// <param name="buffer"></param>
+        /// <param name="cursor"></param>
+        /// <returns></returns>
+        public static bool IsLogigalLineEmpty(this StringBuilder buffer, int cursor)
+        {
+            return 
+                buffer.Length == 0 ||
+                (cursor == buffer.Length && buffer[cursor - 1] == '\n') ||
+                (cursor > 0 && buffer[cursor] == '\n') ||
+                (cursor > 0 && buffer[cursor] == '\n' && cursor < buffer.Length - 1 && buffer[cursor - 1] == '\n')
```
Comment (daxian-dbw) [inline PSReadLine/StringBuilderLinewiseExtensions.cs]: > `(cursor > 0 && buffer[cursor] == '\n' && cursor < buffer.Length - 1 && buffer[cursor - 1] == '\n')`

Is this line really needed? If `cursor > 0 && buffer[cursor] == '\n'` evaluates to `true`, this line will not be evaluated; if that expression evaluates to `false`, then this line will always return false.

File: PSReadLine/StringBuilderTextObjectExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        {
+            // cursor may be past the end of the buffer when calling this method
+            // this may happen if the cursor is at the beginning of a new line
+
+            var i = Math.Min(position, buffer.Length - 1);
+
+            // if starting on a word consider a text object as a sequence of characters excluding the delimiters
+            // otherwise, consider a word as a sequence of delimiters
+
+            var delimiters = wordDelimiters + '\n';
```
Comment (daxian-dbw) [inline PSReadLine/StringBuilderTextObjectExtensions.cs]: Given that `buffer.InWord` excludes white characters, it seems adding `\n` to the delimiters here is useless -- imagine `buffer[i]` is `\n`.

File: PSReadLine/StringBuilderLinewiseExtensions.cs (hunk 1 above)
Comment (springcomp) [reply PSReadLine/StringBuilderLinewiseExtensions.cs]: You are right, this line is redundant.

File: PSReadLine/StringBuilderTextObjectExtensions.cs (hunk 2 above)
Comment (springcomp) [reply PSReadLine/StringBuilderTextObjectExtensions.cs]: For the purpose of this method, the newline `\n` character is considered an additional delimiter from the list of default word delimiters.

In any case, you are correct, it has no effect on the `buffer.InWord` test.
However, is absolutely matters when testing whether the cursor is currently on a delimiter.

I tried to clarify intent by rewriting the code slightly.

https://github.com/springcomp/PSReadLine/blob/68de1c6432dcf8e5e25738c595c0c1ad391c4ef7/PSReadLine/StringBuilderTextObjectExtensions.cs#L25-L45

Please, let me know if that clears things up.

**Review by daxian-dbw — APPROVED** [review]
LGTM

**Comment by sharpchen** [comment]
Can we have it exposed to `[Microsoft.PowerShell.PSConsoleReadLine]` class? I tried to remap this to another chord but failed to find a work around

**Comment by springcomp** [comment]
> Can we have it exposed to `[Microsoft.PowerShell.PSConsoleReadLine]` class? I tried to remap this to another chord but failed to find a work around

Unfortunately, that PR has been merged a bit too quickly.
The current handler is not a public method that can be remapped, unfortunately.

I addressed feedback from follow up conversations in [a future pull request](https://github.com/PowerShell/PSReadLine/pull/3853) that addresses this shortcomings.

**Pull request merged**